New docs section to explain how to setup test environment for Custom Applications (#2587)
commercetools/merchant-center-application-kit: website/src/content/help-needed.mdx

If you are interested in more advanced functionalities, let us know and open a [support issue](https://github.com/commercetools/merchant-center-application-kit/issues/new/choose).

</Info>

> # Using a testing environment

Author: Updated in this commit: 8d77495d8009d3b696b225e937ee439a6b55a490

During the development cycle of new features for a Custom Application, it can be helpful to showcase or test those new features before publishing them to the production environment.

> Merchant Center platform has only one environment where Custom Applications are installed to, but a similar goal can be achieved by setting up the Custom Application configuration so it uses a different [`entryPointUriPath`](/api-reference/application-config#entrypointuripath) configuration property per its installation deployments.

Author: Updated in this commit: 8d77495d8009d3b696b225e937ee439a6b55a490

<Info>

More information about how to set environment specific Custom Applications cofiguration properties can be found in [this section](/api-reference/application-config#environment-variable-references).

</Info>

The very same Custom Application can be deployed several times and set its [configuration file](/api-reference/application-config) to use different `entryPointUriPath` values.

Let's say there are two deployments of the same application: `A` and `B`. We can assign an environment value for the `entryPointUriPath` property in its configuration.

```json
{
  "entryPointUriPath": "${env:ENTRY_POINT_URI_PATH}"
}
```

> The `ENTRY_POINT_URI_PATH` environment variable can be provided in different ways:

Author: I tried to rewrite that section a little bit in this commit: 8d77495d8009d3b696b225e937ee439a6b55a490

However, there is an important update necessary to be applied to the Custom Application source code.
Since we recommend having a constants file where the `entryPointUriPath` should be defined (as it is done is our [starter template](https://github.com/commercetools/merchant-center-application-kit/blob/main/application-templates/starter/src/constants.js#L4)) and that value would be used both at build time (when parsing the [configuration file](https://github.com/commercetools/merchant-center-application-kit/blob/main/application-templates/starter/custom-application-config.mjs#L8)) and at runtime, the initialization of that constant can be static but we need to get it from the environment:

```js highlightLines="4-5"
// Make sure to import the helper functions from the `ssr` entry point.
import { entryPointUriPathToPermissionKeys } from '@commercetools-frontend/application-shell/ssr';

export const entryPointUriPath = typeof window === 'undefined' ?
  process.env.ENTRY_POINT_URI_PATH : window.app.ENTRY_POINT_URI_PATH;

export const PERMISSIONS = entryPointUriPathToPermissionKeys(entryPointUriPath);
```

> The value is read from the environment at build time and from the `window.app` object at runtime (that property is automatically populated by the Custom Application iternals).

Author: I rewrote the section in this commit: 8d77495d8009d3b696b225e937ee439a6b55a490

Please let me know what do you think about it.

</Info>

The very same Custom Application can be deployed several times and set its [configuration file](/api-reference/application-config) to use different `entryPointUriPath` values by taking advantage of environment variables for every deployment.

Let's say there are two deployments of the same application: `A` and `B`. We can assign an environment value for the `entryPointUriPath` property in its configuration.

```json
{
  "entryPointUriPath": "${env:ENTRY_POINT_URI_PATH}"
}
```

The `ENTRY_POINT_URI_PATH` environment variable can be provided in different ways:

For example:
- as an inline environment variable when running a script command

  ```console
  ENTRY_POINT_URI_PATH=my-application-entry-point mc-scripts start
  ```

- using a [dotenv](https://github.com/motdotla/dotenv) file

  ```bash title=".env-A"
  ENTRY_POINT_URI_PATH=my-application-entry-point
  ```

  ```console
  mc-scripts --env .env-A start
  ```

- by defining the environment variables in your CI service

Using this pattern, different `entryPointUriPath` values can be defined for different environments:
* `A` environment: `my-application-entry-point`
* `B` environment: `my-application-entry-point-test`

> With this setup, a new preview version of the Custom Application can be deployed to `B` environment and showcasing/testing can be done without affecting the production environment.

Reviewer: ```suggestion
With this setup, preview versions of the Custom Application can be deployed to the `B` environment and tested using the application `my-application-entry-point-test`.
```

Author: Updated in this commit: 7bdca6330f1f6e6bc799395e5524ae1f4a2b7f9d

> However, there is also another needed update to be applied to the Custom Application source code related to user permissions.
> 
> Every Custom Application has its own pair of permissions (reading and writing) and its names are derived from the configured `entryPointUriPath` (as explained in [this section](/migrating-from-project-level-custom-applications#choose-the-entrypointuripath)).
> Since there will be different entry points by deployment, we need to update the permissions setup accordingly.
> 
> We do recommend setting up a constants file where both the `entryPointUriPath` and the derived permissions are defined as explained in the [permissions section](/custom-applications/development/permissions#defining-constants).<br/>
> Nevertheless is a caveat though since the `entryPointUriPath` is not a static value in this scenario so we would need to tailor the source code to calculate it accordingly as we showcase in the next example.
> 
> ```js highlightLines="4-5"
> // Make sure to import the helper functions from the `ssr` entry point.
> import { entryPointUriPathToPermissionKeys } from '@commercetools-frontend/application-shell/ssr';
> 
> export const entryPointUriPath = typeof window === 'undefined' ?
>   process.env.ENTRY_POINT_URI_PATH : window.app.ENTRY_POINT_URI_PATH;
> 
> export const PERMISSIONS = entryPointUriPathToPermissionKeys(entryPointUriPath);
> ```
> 
> The value is read from the environment at build time and from the `window.app` object at runtime (that property is automatically populated by the Custom Application iternals).

Reviewer: ````suggestion
However, besides using different `entryPointUriPath` values, the logic of verifying the user permissions in the Custom Application source code must also be adjusted accordingly.

As [explained in this section](/migrating-from-project-level-custom-applications#choose-the-entrypointuripath), the permission names of a Custom Application are derived from the `entryPointUriPath` value. Therefore, using different `entryPointUriPath` values require to change the permission names as well.

We recommend to use the helper function `entryPointUriPathToPermissionKeys` to compute the permission names from the `entryPointUriPath`.

For more information see the [Permissions](/development/permissions) documentation.

<Info>

In this scenario the `entryPointUriPath` value cannot be defined statically but it's injected using an environment variable.

Therefore, reading the value from the source code must take into account both server-side and client-side logic.

```js highlightLines="4-6"
// Make sure to import the helper functions from the `ssr` entry point.
import { entryPointUriPathToPermissionKeys } from '@commercetools-frontend/application-shell/ssr';

export const entryPointUriPath = typeof window === 'undefined'
  ? process.env.ENTRY_POINT_URI_PATH
  : window.app.ENTRY_POINT_URI_PATH;

export const PERMISSIONS = entryPointUriPathToPermissionKeys(entryPointUriPath);
```

</Info>
````

Author: Updated in this commit: 7bdca6330f1f6e6bc799395e5524ae1f4a2b7f9d

Thanks a lot for your help 🙇

If you are interested in more advanced functionalities, let us know and open a [support issue](https://github.com/commercetools/merchant-center-application-kit/issues/new/choose).

</Info>

# I want to use a test or staging environment for my Custom Application

During the development cycle of new features for a Custom Application, it can be helpful to showcase or test those new features before publishing them to the production environment.

This generally can be achieved using preview deployments, where you can inspect different versions of the Custom Application. As of now, the Merchant Center does not support this feature yet.

The only option to emulate this setup is to create multiple Custom Applications in the Merchant Center, each one with a different [`entryPointUriPath`](/api-reference/application-config#entrypointuripath).

> However, the Custom Application must be configured to support "dynamic" values for the `entryPointUriPath` and build different versions of it.

Reviewer 2: ```suggestion
However, the Custom Application must be configured to support dynamic values for the `entryPointUriPath` and build different versions of it.
```

<Info>

> Custom Applications "dynamic" configurations (per environment) setup are further explained in [this section](/api-reference/application-config#environment-variable-references).

Author: I was wondering if something like this could also work:
```
Setting up dynamic configurations of Custom Applications (per environment) is further...
```
